From 7fd651b0b712647b505ae02b3486eb62395738ee Mon Sep 17 00:00:00 2001
From: Myron Stowe <myron.stowe@redhat.com>
Date: Sat, 19 Jul 2014 04:29:27 -0400
Subject: [platform] x86: Use global PCI rescan-remove locking

Message-id: <20140719042926.19438.94877.stgit@gir.stowe>
Patchwork-id: 86026
O-Subject: [RHEL7.1 PATCH 64/81] platform / x86: Use global PCI rescan-remove locking
Bugzilla: 1110895
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1110895
Upstream Status: 8b9ec1da6a310b438516468c669a7e1be046b18c

commit 8b9ec1da6a310b438516468c669a7e1be046b18c
Author: Rafael J. Wysocki <rafael.j.wysocki@intel.com>
Date:   Fri Jan 10 15:27:08 2014 +0100

    platform / x86: Use global PCI rescan-remove locking

    Multiple race conditions are possible between the rfkill hotplug in the
    asus-wmi and eeepc-laptop drivers and the generic PCI bus rescan and device
    removal that can be triggered via sysfs.

    To avoid those race conditions make asus-wmi and eeepc-laptop use global
    PCI rescan-remove locking around the rfkill hotplug.

    Signed-off-by: Rafael J. Wysocki <rafael.j.wysocki@intel.com>
    Signed-off-by: Bjorn Helgaas <bhelgaas@google.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/platform/x86/asus-wmi.c b/drivers/platform/x86/asus-wmi.c
index 21889db..7682f64 100644
--- a/drivers/platform/x86/asus-wmi.c
+++ b/drivers/platform/x86/asus-wmi.c
@@ -607,6 +607,7 @@ static void asus_rfkill_hotplug(struct asus_wmi *asus)
  mutex_unlock(&asus->wmi_lock);
 
  mutex_lock(&asus->hotplug_lock);
+ pci_lock_rescan_remove();
 
  if (asus->wlan.rfkill)
   rfkill_set_sw_state(asus->wlan.rfkill, blocked);
@@ -657,6 +658,7 @@ static void asus_rfkill_hotplug(struct asus_wmi *asus)
  }
 
 out_unlock:
+ pci_unlock_rescan_remove();
  mutex_unlock(&asus->hotplug_lock);
 }
 
diff --git a/drivers/platform/x86/eeepc-laptop.c b/drivers/platform/x86/eeepc-laptop.c
index 5d26e70..4d2b1a5 100644
--- a/drivers/platform/x86/eeepc-laptop.c
+++ b/drivers/platform/x86/eeepc-laptop.c
@@ -598,6 +598,7 @@ static void eeepc_rfkill_hotplug(struct eeepc_laptop *eeepc, acpi_handle handle)
   rfkill_set_sw_state(eeepc->wlan_rfkill, blocked);
 
  mutex_lock(&eeepc->hotplug_lock);
+ pci_lock_rescan_remove();
 
  if (eeepc->hotplug_slot) {
   port = acpi_get_pci_dev(handle);
@@ -655,6 +656,7 @@ out_put_dev:
  }
 
 out_unlock:
+ pci_unlock_rescan_remove();
  mutex_unlock(&eeepc->hotplug_lock);
 }
 
-- 
1.7.1